File: 200-number-of-islands/number-of-islands.java
/*
<Idea>
 - Pick a certain land(vertex) and do BFS to figure out the constituent of island that vertex belongs to.
 - Pick another vertex which has not been visited. Do BFS from there and go on.

<Time Complexity>
 - O(m*n) (Since 1 <= m, n <= 300, It would be efficient enough.)
<Concept>
 - Counting the nubmer of connected components.
*/


class Solution {
    public int numIslands(char[][] grid) {
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        int numberOfIslands = 0;

        for(int row = 0; row < grid.length; row++){
            for(int col = 0; col < grid[0].length; col++){
                if(grid[row][col] == '1' && !visited[row][col]){
                    BFS(grid, visited, row, col);
                    numberOfIslands++;
                }
            }
        }

        return numberOfIslands;
    }

    private void BFS(char[][] grid, boolean[][] visited, int row, int col){
        ArrayDeque<int[]> q = new ArrayDeque<>();
        
        int[] dRow = new int[]{-1, 1, 0, 0};
        int[] dCol = new int[]{0, 0, -1, 1};

        visited[row][col] = true;
        q.addLast(new int[]{row, col});

        while(!q.isEmpty()){
            int[] current = q.removeFirst();
            for(int d = 0; d < 4; d++){
                int adjRow = current[0] + dRow[d];
                int adjCol = current[1] + dCol[d];
                if(isAvailable(grid.length, grid[0].length, adjRow, adjCol)
                 && grid[adjRow][adjCol] == '1'
                  && !visited[adjRow][adjCol]){
                    visited[adjRow][adjCol] = true;
                    q.addLast(new int[]{adjRow, adjCol});
                }
            }
        }
    }

    private boolean isAvailable(int rowLen, int colLen, int row, int col){
        return row >= 0 && row < rowLen && col >= 0 && col < colLen;
    }
}

File: 200-number-of-islands/number-of-islands.kt
import java.util.ArrayDeque

class Solution {

    fun numIslands(grid: Array<CharArray>): Int {
        val n = grid.size
        val m = grid[0].size
        val dRow = intArrayOf(-1, 1, 0, 0)
        val dCol = intArrayOf(0, 0, -1, 1)

        var islandCnt = 0
        val visited = Array(n) { BooleanArray(m) }

        for(r in 0..<n){for(c in 0..<m) {

            if(grid[r][c] == '1' && !visited[r][c]){
                val q = ArrayDeque<IntArray>()

                q.offer(intArrayOf(r, c))
                visited[r][c] = true

                while(!q.isEmpty()) {
                    val current = q.poll()

                    for(d in 0..3){
                        val nextRow = current[0] + dRow[d]
                        val nextCol = current[1] + dCol[d]

                        if(isAvailable(nextRow, nextCol, n, m)
                         && grid[nextRow][nextCol] == '1'
                          && !visited[nextRow][nextCol])
                          {
                            visited[nextRow][nextCol] = true
                            q.offer(intArrayOf(nextRow, nextCol))
                          }
                    }
                }

                islandCnt++
            }
        }}

        
        return islandCnt
    }

    fun isAvailable(row: Int, col: Int, n: Int, m: Int): Boolean {
        return row >= 0 && row < n && col >= 0 && col < m
    }
}